### PR TITLE
Fix #2538 and #2801

### LINKED ISSUE #2538: `export default class` throws reference error when the class name is referenced inside the declaration
This compiles and the output is correct.

```typescript
class One {
    foo(): One {
        return this;
    }
}

export default One;
```

```javascript
var One = (function () {
    function One() {
    }
    One.prototype.foo = function () {
        return this;
    };
    return One;
})();
exports.default = One;
```

---

Combining the class declaration and export doesn't compile:

```typescript
export default class One {
    foo(): One {
        return this;
    }
}
```

```
module1.ts(2,12): error TS2304: Cannot find name 'One'.
```

### PR DESCRIPTION
**Pull request by mhegazy**
#2538: Missed check for ambient module declaration in resolving default class name
#2801: Clean up the module preamble writing code to ensure modules with no aliases still result in valid JS

### CODE REVIEW

File: src/compiler/diagnosticInformationMap.generated.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // <auto-generated />
 /// <reference path="types.ts" />
+/* @internal */
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticInformationMap.generated.ts]: This is just the worst.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -349,7 +349,8 @@ module ts {
                             }
                             result = undefined;
                         }
-                        else if (location.kind === SyntaxKind.SourceFile) {
+                        else if (location.kind === SyntaxKind.SourceFile ||
+                            (location.kind === SyntaxKind.ModuleDeclaration && (<ModuleDeclaration>location).name.kind === SyntaxKind.StringLiteral)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Would `hasExternalModuleSymbol` help you out at all?

I notice we do this `name.kind === SyntaxKind.StringLiteral` thing a lot. Can we abstract this to a helper function?

Comment (mhegazy) [inline src/compiler/checker.ts]: We should. will get that in a different commit.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -4649,44 +4636,78 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
+                /// The location of the alias in the parameter list in the factory function needs to 
+                /// match the position of the module name in the dependency list.
+                ///
+                /// To ensure this is true in cases of modules with no aliases, e.g.: 
+                /// `import "module"` or `<amd-dependency path= "a.css" \>` we need to add these 
+                /// modules to the end of the dependencies list
+
+                let aliasedModuleNames: string[] = [];
+                let unaliasedModuleNames: string[] = [];
+                let importAliaseNames: string[] = [];
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `importAliaseNames` -> `importAliasNames`

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -4649,44 +4636,78 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
+                        unaliasedModuleNames.push(externalModuleName);
                     }
                 }
+
+                // Fill in amd-dependency tags
                 for (let amdDependency of node.amdDependencies) {
                     if (amdDependency.name) {
-                        write(", ");
-                        write(amdDependency.name);
+                        aliasedModuleNames.push("\"" + amdDependency.path + "\"");
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Let's just use single quoted strings

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -4649,44 +4636,78 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
-                    let text = "\"" + amdDependency.path + "\"";
-                    write(", ");
-                    write(text);
-                }
-                write("], function (require, exports");
-                for (let importNode of externalImports) {
-                    write(", ");
+
+                    // Find the name of the module alais, if there is one
+                    var importAliasName: string;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `let`

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -4649,44 +4636,78 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
                 collectExternalModuleInfo(node);
-                writeLine();
-                write("define(");
-                sortAMDModules(node.amdDependencies);
-                if (node.amdModuleName) {
-                    write("\"" + node.amdModuleName + "\", ");
-                }
-                write("[\"require\", \"exports\"");
+                
+                /// An AMD define function has the following shape:
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Single quotes around 'define'

File: src/compiler/emitter.ts (hunk 4 above)
Comment (mhegazy) [inline src/compiler/emitter.ts]: we have not used them before.. i really do not feel strongly either ways but i would keep them for consistency.

File: src/compiler/emitter.ts (hunk 5 above)
Comment (mhegazy) [inline src/compiler/emitter.ts]: done.

File: src/compiler/emitter.ts [hunk 7]
```diff
@@ -4649,44 +4636,78 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
+                ///     define(id?, dependencies?, factory);
+                ///
+                /// This has the shape of
+                ///     define(name, ["module1", "module2"], function (module1Alias) {
+                /// The location of the alias in the parameter list in the factory function needs to 
+                /// match the position of the module name in the dependency list.
+                ///
+                /// To ensure this is true in cases of modules with no aliases, e.g.: 
+                /// `import "module"` or `<amd-dependency path= "a.css" \>` we need to add these 
+                /// modules to the end of the dependencies list
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: "to add these" -> "to add modules without alias names to the end"

Also note here that `aliasedModuleNames` and `importAliasNames` need to have corresponding indices for their module names and aliasing names

**Comment by DanielRosenwasser** [comment]
:+1: after comments are addressed.

**Pull request merged**